src/ray/core_worker/reference_count.h

```diff
@@ -3,18 +3,95 @@
 class ReferenceCounter {
  public:
-  ReferenceCounter() {}
+  /// Metadata for an ObjectID reference in the language frontend.
+  struct Reference {
+    /// Constructor for a reference whose origin is unknown.
+    Reference() : owned_by_us(false) {}
+    /// Constructor for a reference that we created.
+    Reference(const TaskID &owner_id, const rpc::Address &owner_address)
+        : owned_by_us(true), owner({owner_id, owner_address}) {}
+
+    size_t RefCount() const {
+      return local_ref_count + submitted_task_ref_count + contained_in_owned.size() +
+                     contained_in_borrowed_id.has_value()
+                 ? 1
+                 : 0;
+    }
+
+    bool IsBorrower() const {
+      return local_ref_count + submitted_task_ref_count + contained_in_owned.size() > 0;
+    }
+
+    size_t NumBorrowers() const { return borrowers.size(); }
+
+    rpc::ObjectReferenceCount ToProto() const {
+      rpc::ObjectReferenceCount ref;
+      if (owner.has_value()) {
+        ref.mutable_reference()->set_owner_id(owner->first.Binary());
+        ref.mutable_reference()->mutable_owner_address()->CopyFrom(owner->second);
+      }
+      bool has_local_ref = RefCount() > 0;
+      ref.set_has_local_ref(has_local_ref);
+      for (const auto &borrower : borrowers) {
+        ref.add_borrowers()->CopyFrom(borrower.ToProto());
+      }
+      if (contained_in_borrowed_id.has_value()) {
+        ref.set_contained_in_borrowed_id(contained_in_borrowed_id->Binary());
+      }
+      for (const auto &contains_id : contains) {
+        ref.add_contains(contains_id.Binary());
+      }
+      return ref;
+    }
+
+    /// The local ref count for the ObjectID in the language frontend.
+    size_t local_ref_count = 0;
+    /// The ref count for submitted tasks that depend on the ObjectID.
+    size_t submitted_task_ref_count = 0;
+    /// Whether we own the object. If we own the object, then we are
+    /// responsible for tracking the state of the task that creates the object
+    /// (see task_manager.h).
+    bool owned_by_us;
+    /// The object's owner, if we know it. This has no value if the object is
+    /// if we do not know the object's owner (because distributed ref counting
+    /// is not yet implemented).
+    absl::optional<std::pair<TaskID, rpc::Address>> owner;
+    /// Callback that will be called when this ObjectID no longer has references.
+    std::function<void(const ObjectID &)> on_delete;
+
+    /// Callback for a borrower that is called when this process is no longer a
+    /// borrower.
+    std::function<void()> on_local_ref_deleted;
+    /// Object IDs that contain this object ID. This field contains only object
+    /// IDs that we own.
+    absl::flat_hash_set<ObjectID> contained_in_owned;
+    /// Object ID that contain this object ID. Needed for nested Object IDs.
+    /// This field must be an ObjectID that we are currently borrowing.
+    absl::optional<ObjectID> contained_in_borrowed_id;
```

Comment:
Please document the semantics of these (mutable) fields, including the situations when they are updated.